contracts/5/IbETHRouter.sol

```diff
@@ -0,0 +1,302 @@
+        TransferHelper.safeTransferFrom(alpha, msg.sender, address(this), amountAlphaInMax);
+        address[] memory path = new address[](2);
+        path[0] = alpha;
+        path[1] = ibETH;
+        IBank(ibETH).withdraw(0);
+        uint256[] memory swapAmounts = IUniswapV2Router02(router).swapTokensForExactTokens(ibETHForExactETH(amountETHOut), amountAlphaInMax, path, address(this), deadline);                           
+        IBank(ibETH).withdraw(swapAmounts[1]);
+        amounts = new uint256[](2);
+        amounts[0] = swapAmounts[0];
+        amounts[1] = address(this).balance;
+        TransferHelper.safeTransferETH(to, address(this).balance);        
+        if (amountAlphaInMax > amounts[0]) {
+            TransferHelper.safeTransfer(alpha, msg.sender, amountAlphaInMax.sub(amounts[0]));
+        }                    
+    }
+
+    // Swap exact amount of Token for ETH
+    // 1. Receive Token from caller
+    // 2. Swap Token for ibETH.
+    // 3. Unwrap ibETH to ETH.
+    function swapExactAlphaForETH(
+        uint256 amountAlphaIn,
+        uint256 amountETHOutMin,         
+        address to,
+        uint256 deadline
+    ) external returns (uint256[] memory amounts) {
+        TransferHelper.safeTransferFrom(alpha, msg.sender, address(this), amountAlphaIn); 
+        address[] memory path = new address[](2);
+        path[0] = alpha;
+        path[1] = ibETH;
+        uint256[] memory swapAmounts = IUniswapV2Router02(router).swapExactTokensForTokens(amountAlphaIn, 0, path, address(this), deadline);                        
+        IBank(ibETH).withdraw(swapAmounts[1]);        
+        amounts = new uint256[](2);
+        amounts[0] = swapAmounts[0];
+        amounts[1] = address(this).balance;
+        TransferHelper.safeTransferETH(to, amounts[1]);
+        require(amounts[1] >= amountETHOutMin, "IbETHRouter: receive less ETH than amountETHmin");                                       
+    }
+
+    // Swap ETH for exact amount of Token
+    // 1. Receive ETH from caller
+    // 2. Wrap ETH to ibETH.
+    // 3. Swap ibETH for Token    
+    function swapETHForExactAlpha(
+        uint256 amountAlphaOut,          
+        address to,
+        uint256 deadline
+    ) external payable returns (uint256[] memory amounts) {             
+        IBank(ibETH).deposit.value(msg.value)();              
+        uint256 amountIbETHInMax = IBank(ibETH).balanceOf(address(this));        
+        address[] memory path = new address[](2);
+        path[0] = ibETH;
+        path[1] = alpha;                
+        uint256[] memory swapAmounts = IUniswapV2Router02(router).swapTokensForExactTokens(amountAlphaOut, amountIbETHInMax, path, to, deadline);                                                
+        amounts = new uint256[](2);               
+        amounts[1] = swapAmounts[1];
+        // Transfer left over ETH back
+        if (amountIbETHInMax > swapAmounts[0]) {                         
```

Comment:
`amounts[0]` is not set if the if statement does not hit. Is this expected?

Reply:
Nope. My misunderstanding here.

1.  `amountIbETHInMax < swapAmounts[0]`
This case will revert in UniswapRouter.
2. `amountIbETHInMax > swapAmounts[0]`
This case will return dust ETH to user, so `amounts[0] = msg.value - address(this).balance`.
3. `amountIbETHInMax = swapAmounts[0]`
This case will use all ETH caller provided, so `amounts[0] = msg.value` .